tests/core/remodel-base/task.yaml

```diff
@@ -12,15 +12,33 @@ prepare: |
     . "$TESTSLIB"/core-config.sh
     #shellcheck source=tests/lib/systemd.sh
     . "$TESTSLIB"/systemd.sh
-    systemctl stop snapd.service snapd.socket
 
-    clean_snapd_lib
-    prepare_core_model
-    prepare_test_account valid-for-testing
-    prepare_test_model valid-for-testing-pc
+    if [ "$SPREAD_REBOOT" = 0 ]; then
+        REBOOT_NEEDED=false
+        if os.query is-core18 && snap list snapd | NOMATCH ' x1 '; then
+            REBOOT_NEEDED=true
+        fi
+        systemctl stop snapd.service snapd.socket
+
+        clean_snapd_lib
+        prepare_core_model
+        prepare_test_account valid-for-testing
+        prepare_test_model valid-for-testing-pc
+
+        # kick first boot again
+        systemctl start snapd.service snapd.socket
 
-    # kick first boot again
-    systemctl start snapd.service snapd.socket
+        if [ "$REBOOT_NEEDED" = true ]; then
```

Comment:
Maybe not too test friendly, but don't need to start snapd right above, but rather just reboot right away after preparing the model. Should work in both scenarios and at least we know that the system is in a clean state?

so we'd have:
```
  if [ "$SPREAD_REBOOT" = 0 ]; then
        systemctl stop snapd.service snapd.socket
        clean_snapd_lib
        prepare_core_model
        prepare_test_account valid-for-testing
        prepare_test_model valid-for-testing-pc
        # kick first boot again
        REBOOT
fi
```
and similarly in restore?

Reply:
@bboozzoo, thanks for the review tried this but the test is failing with a similar error than in the original test

$ spread -debug external:ubuntu-core-18-64:tests/core/remodel
2022-02-07 11:43:23 Project content is packed for delivery (8.75MB).
2022-02-07 11:43:23 If killed, discard servers with: spread -reuse-pid=338819 -discard
2022-02-07 11:43:23 Allocating external:ubuntu-core-18-64...
2022-02-07 11:43:23 Waiting for external:ubuntu-core-18-64 to make SSH available at localhost:8022...
2022-02-07 11:43:23 Allocated external:ubuntu-core-18-64.
2022-02-07 11:43:23 Connecting to external:ubuntu-core-18-64...
2022-02-07 11:43:24 Connected to external:ubuntu-core-18-64 at localhost:8022.
2022-02-07 11:43:24 Sending project content to external:ubuntu-core-18-64...
2022-02-07 11:43:24 Preparing external:ubuntu-core-18-64 (external:ubuntu-core-18-64)...
2022-02-07 11:43:24 Preparing external:ubuntu-core-18-64:tests/core/ (external:ubuntu-core-18-64)...
2022-02-07 11:45:17 Preparing external:ubuntu-core-18-64:tests/core/remodel (external:ubuntu-core-18-64)...
2022-02-07 11:45:18 Rebooting on external:ubuntu-core-18-64 (external:ubuntu-core-18-64:tests/core/remodel) as requested...
2022-02-07 11:50:42 WARNING: external:ubuntu-core-18-64 (external:ubuntu-core-18-64:tests/core/remodel) running late. Current output:
-----
(... 76 lines above ...)
tests.invariant: leftover-defer-sh ok
error: cannot list snaps: cannot communicate with server: Get http://localhost/v2/snaps?select=all: dial unix /run/snapd.socket: connect: connection refused
tests.invariant: broken-snaps ok
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/core-config.sh
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/systemd.sh
+ '[' 1 = 0 ']'
+ wait_for_first_boot_change
+ retry -n 200 --wait 1 sh -c 'snap changes | MATCH "Done.*Initialize system state"'
error: cannot communicate with server: Get http://localhost/v2/changes?select=all: dial unix /run/snapd.socket: connect: connection refused
grep error: pattern not found, got:
-----

Reply:
@bboozzoo well, what I see is that the system is installing the snapd and kernel after we start snapd. This triggers a reboot which I force with the REBOOT. In case of the regular spread tests, it is happening that it can't reinstall snapd because it is unasserted, so the test works well without the REBOOT.

When, I tried what you suggested, the reboot happens but the after reboot, snapd is not responding anymore, even I can't see errors in the journal log and also the snapd service is running.

Reply:
Something interesting is that the reboot is triggered on uc18 but in uc16 is not.

Reply:
Ok, let's keep an eye on this. I suspect there's still something fishy about asserted vs unasserted snaps